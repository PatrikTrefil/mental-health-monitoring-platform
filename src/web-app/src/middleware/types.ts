import { NextMiddlewareWithAuth } from "next-auth/middleware";

export type MiddlewareWrapper = (
    middleware: NextMiddlewareWithAuth
) => NextMiddlewareWithAuth;
