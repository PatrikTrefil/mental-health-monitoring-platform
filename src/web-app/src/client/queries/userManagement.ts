import { createQueryKeys } from "@lukemorales/query-key-factory";
import { SortingState } from "@tanstack/react-table";
import {
    loadClientPatient,
    loadClientsAndPatients,
    loadRoles,
    loadSpravceDotazniku,
    loadZadavatelDotazniku,
} from "../userManagementClient";

export const usersQuery = createQueryKeys("users", {
    list: ({
        formioToken,
        pagination,
        sort,
        filters,
    }: {
        formioToken: string;
        pagination: { limit: number; offset: number };
        sort?: { field: string; order: "asc" | "desc" };
        filters?: {
            fieldPath: string;
            operation: "contains";
            comparedValue: string;
        }[];
    }) => ({
        // We don't include the token in the query key, because the result does not depend on it
        // eslint-disable-next-line @tanstack/query/exhaustive-deps
        queryKey: ["usersList", pagination, sort, filters],
        queryFn: () =>
            loadClientsAndPatients({
                token: formioToken,
                pagination,
                sort,
                filters,
            }),
    }),
    detail: (formioToken: string, userSubmissionId: string) => ({
        // We don't include the token in the query key, because the result does not depend on it
        // eslint-disable-next-line @tanstack/query/exhaustive-deps
        queryKey: [userSubmissionId],
        queryFn: () => loadClientPatient(userSubmissionId, formioToken),
    }),
});

export const zadavatelDotaznikuQuery = createQueryKeys("zadavatelDotazniku", {
    list: ({
        formioToken,
        pagination,
        sort,
    }: {
        formioToken: string;
        pagination: { limit: number; offset: number };
        sort?: { field: string; order: "asc" | "desc" };
    }) => ({
        // We don't include the token in the query key, because the result does not depend on it
        // eslint-disable-next-line @tanstack/query/exhaustive-deps
        queryKey: ["zadavatelDotaznikuList", pagination, sort],
        queryFn: () =>
            loadZadavatelDotazniku({ token: formioToken, pagination, sort }),
    }),
});

export const spravceDotaznikuQuery = createQueryKeys("spravceDotazniku", {
    list: ({
        formioToken,
        pagination,
        sort,
    }: {
        formioToken: string;
        pagination: { limit: number; offset: number };
        sort?: { field: string; order: "asc" | "desc" };
    }) => ({
        // We don't include the token in the query key, because the result does not depend on it
        // eslint-disable-next-line @tanstack/query/exhaustive-deps
        queryKey: ["spravceDotaznikuList", pagination, sort],
        queryFn: () =>
            loadSpravceDotazniku({ token: formioToken, pagination, sort }),
    }),
});

export const rolesQuery = createQueryKeys("roles", {
    list: (formioToken: string) => ({
        // We don't include the token in the query key, because the result does not depend on it
        // eslint-disable-next-line @tanstack/query/exhaustive-deps
        queryKey: ["roles"], // Needs to be non-empty
        queryFn: () => loadRoles(formioToken),
    }),
});

export const employeesInfiniteQuery = createQueryKeys("employees", {
    list: ({
        formioToken,
        pageSize,
        sorting,
        filters,
    }: {
        formioToken: string;
        pageSize: number;
        sorting: SortingState;
        filters?: {
            fieldPath: string;
            operation: "contains";
            comparedValue: string;
        }[];
    }) => ({
        queryFn: async ({ pageParam }) => {
            /**
             * If the offset is undefined, there are no more pages to load.
             */
            let offsetSpravce: number | undefined;
            /**
             * If the offset is undefined, there are no more pages to load.
             */
            let offsetZadavatel: number | undefined;

            // When the first query is made, the param is not present
            const isFirstPage = pageParam === undefined;
            if (isFirstPage) {
                offsetSpravce = 0;
                offsetZadavatel = 0;
            } else {
                offsetSpravce = pageParam.nextPageSpravceOffset;
                offsetZadavatel = pageParam.nextPageZadavatelOffset;
            }
            const spravciPromise =
                offsetSpravce !== undefined
                    ? loadSpravceDotazniku({
                          token: formioToken,
                          pagination: {
                              limit: pageSize,
                              offset: offsetSpravce,
                          },
                          sort:
                              sorting[0] !== undefined
                                  ? {
                                        field: sorting[0].id,
                                        order: sorting[0].desc ? "desc" : "asc",
                                    }
                                  : undefined,
                          filters,
                      })
                    : new Promise<undefined>((resolve) => resolve(undefined));
            const zadavatelePromise =
                offsetZadavatel !== undefined
                    ? loadZadavatelDotazniku({
                          token: formioToken,
                          pagination: {
                              limit: pageSize,
                              offset: offsetZadavatel,
                          },
                          sort:
                              sorting[0] !== undefined
                                  ? {
                                        field: sorting[0].id,
                                        order: sorting[0].desc ? "desc" : "asc",
                                    }
                                  : undefined,
                          filters,
                      })
                    : new Promise<undefined>((resolve) => resolve(undefined));

            const [dataSpravceDotazniku, dataZadavatelDotazniku] =
                await Promise.all([spravciPromise, zadavatelePromise]);

            const {
                resData,
                nextPageOffsetA: nextPageSpravceOffset,
                nextPageOffsetB: nextPageZadavatelOffset,
            } = mergeSortedPaginatedData(
                dataSpravceDotazniku,
                dataZadavatelDotazniku,
                {
                    pageSize,
                    sorting,
                    offsetA: offsetSpravce,
                    offsetB: offsetZadavatel,
                }
            );

            return {
                data: resData,
                nextPageParam: {
                    nextPageSpravceOffset,
                    nextPageZadavatelOffset,
                },
            };
        },
        // eslint-disable-next-line @tanstack/query/exhaustive-deps
        queryKey: ["employees", sorting, filters, pageSize],
    }),
});

/**
 * Merge sorted data from two sources and calculate offsets for the next page.
 * @param dataA - The first source of data.
 * @param dataB - The second source of data.
 * @param root0 - Other parameters.
 * @param root0.pageSize - The size of a single page of data.
 * @param root0.sorting - Sorting configuration.
 * @param root0.offsetA - The offset of the first source of data.
 * @param root0.offsetB - The offset of the second source of data.
 */
function mergeSortedPaginatedData<TData>(
    dataA: { data: TData[]; totalCount: number } | undefined,
    dataB: { data: TData[]; totalCount: number } | undefined,
    {
        pageSize,
        sorting,
        offsetA,
        offsetB,
    }: {
        pageSize: number;
        sorting: SortingState;
        offsetA: number | undefined;
        offsetB: number | undefined;
    }
): {
    resData: TData[];
    nextPageOffsetA: number | undefined;
    nextPageOffsetB: number | undefined;
} {
    console.debug("Merging data...", {
        dataSpravceDotazniku: dataA,
        dataZadavatelDotazniku: dataB,
        sorting,
        pageSize,
    });
    let resData: TData[] = [];

    /**
     * If the offset is undefined, there are no more pages to load.
     */
    let nextPageOffsetA: number | undefined;
    /**
     * If the offset is undefined, there are no more pages to load.
     */
    let nextPageOffsetB: number | undefined;

    if (dataA === undefined && dataB !== undefined) {
        resData = dataB.data;
        const wouldBeNextPageOffsetB = offsetB! + dataB.data.length;
        nextPageOffsetB =
            wouldBeNextPageOffsetB < dataB.totalCount
                ? wouldBeNextPageOffsetB
                : undefined;
    } else if (dataA !== undefined && dataB === undefined) {
        resData = dataA.data;
        const wouldBeNextPageOffsetA = offsetA! + dataA.data.length;
        nextPageOffsetA =
            wouldBeNextPageOffsetA < dataA.totalCount
                ? wouldBeNextPageOffsetA
                : undefined;
    } else if (dataA !== undefined && dataB !== undefined) {
        let indexDataA = 0;
        let indexDataB = 0;
        while (resData.length < pageSize) {
            const isDataAExhausted = indexDataA >= dataA.data.length;
            if (isDataAExhausted) {
                for (; indexDataB < dataB.data.length; indexDataB++) {
                    resData.push(dataB.data[indexDataB]!);
                }
                break;
            }
            const isDataBExhausted = indexDataB >= dataB.data.length;
            if (isDataBExhausted) {
                for (; indexDataA < dataA.data.length; indexDataA++) {
                    resData.push(dataA.data[indexDataA]!);
                }
                break;
            }
            if (sorting[0] !== undefined) {
                let dataAValue = accessNestedValue(
                    dataA.data[indexDataA],
                    sorting[0].id
                );
                let dataBValue = accessNestedValue(
                    dataB.data[indexDataB],
                    sorting[0].id
                );
                if (sorting[0].desc)
                    resData.push(
                        (dataAValue as any) > (dataBValue as any)
                            ? dataA.data[indexDataA++]!
                            : dataB.data[indexDataB++]!
                    );
                else
                    resData.push(
                        (dataAValue as any) < (dataBValue as any)
                            ? dataA.data[indexDataA++]!
                            : dataB.data[indexDataB++]!
                    );
            } else {
                resData.push(dataA.data[indexDataA++]!);
            }
        }
        const wouldBeNextPageOffsetA = offsetA! + indexDataA;
        const wouldBeNextPageOffsetB = offsetB! + indexDataB;
        nextPageOffsetA =
            wouldBeNextPageOffsetA < dataA.totalCount
                ? wouldBeNextPageOffsetA
                : undefined;
        nextPageOffsetB =
            wouldBeNextPageOffsetB < dataB.totalCount
                ? wouldBeNextPageOffsetB
                : undefined;
    }

    console.debug("Merged data", {
        data: resData,
        nextPageParam: {
            nextPageSpravceOffset: nextPageOffsetA,
            nextPageZadavatelOffset: nextPageOffsetB,
        },
    });

    return {
        resData,
        nextPageOffsetA: nextPageOffsetA,
        nextPageOffsetB: nextPageOffsetB,
    };
}

/**
 * Access a nested value in an object based on string path.
 * @param obj - The object to extract the value from.
 * @param path - The path to the value, e.g. "a.b.c".
 */
function accessNestedValue(obj: unknown, path: string): unknown {
    let value = obj as any;
    for (const key of path.split(".")) {
        if (typeof value === "object" && value !== null && key in value)
            value = value[key];
        else {
            value = null;
            break;
        }
    }
    return value;
}
