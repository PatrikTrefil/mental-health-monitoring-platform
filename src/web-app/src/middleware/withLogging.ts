import { MiddlewareWrapper } from "./types";

/**
 * Middleware for logging requests in format: "METHOD PATHNAME".
 * @param middlewareToWrap - Middleware to wrap around.
 */
const withLogging: MiddlewareWrapper = (middlewareToWrap) => {
    return async (request, _next) => {
        console.debug(request.method, request.nextUrl.pathname);
        return middlewareToWrap(request, _next);
    };
};

export default withLogging;
