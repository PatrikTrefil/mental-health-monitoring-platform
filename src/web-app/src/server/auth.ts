import {
    RequestError,
    fetchRoleList,
    loginAdmin,
    loginUser,
    refreshToken,
} from "@/client/formioClient";
import { UserFormSubmission } from "@/types/userFormSubmission";
import { UserRoleTitle, UserRoleTitles } from "@/types/users";
import { type GetServerSidePropsContext } from "next";
import { AuthOptions, DefaultSession, getServerSession } from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import { z } from "zod";

// make auto-completion work for useSession
declare module "next-auth" {
    interface User extends UserFormSubmission {
        roleTitles: UserRoleTitle[];
        formioToken: string;
        formioTokenExpiration: number;
    }
    interface Session {
        user: UserFormSubmission &
            DefaultSession["user"] & {
                roleTitles: UserRoleTitle[];
                formioToken: string;
            };
    }
}

declare module "next-auth/jwt" {
    interface JWT {
        user?: UserFormSubmission &
            DefaultSession["user"] & {
                roleTitles: UserRoleTitle[];
                formioToken: string;
                formioTokenExpiration: number;
            };
    }
}

/**
 * In miliseconds
 */
const formioTokenExpirationTime = 5 * 60 * 1000; // 5 minutes (default in formio)

export const authOptions: AuthOptions = {
    // Configure one or more authentication providers
    providers: [
        CredentialsProvider({
            name: "Credentials",
            /**
             * Used to create a login page and used as a param of the authorize function
             */
            credentials: {
                ID: { label: "ID", type: "text", placeholder: "1234" },
                password: { label: "Password", type: "password" },
            },
            /**
             * Called when initializing a session. The returned object will be passed to the jwt callback.
             * @returns Promise<null> if the credentials are invalid, otherwise Promise<User>
             */
            async authorize(credentials) {
                if (!credentials) return null; // invalid credentials because they were not provided

                try {
                    const { user, formioToken } = await loginUser(
                        credentials.ID,
                        credentials.password
                    );

                    const adminToken = await loginAdmin(
                        process.env.FORMIO_ROOT_EMAIL,
                        process.env.FORMIO_ROOT_PASSWORD
                    );

                    const roleList = await fetchRoleList(adminToken);

                    const roleTitles = user.roles.map((roleId) => {
                        const title = roleList.find(
                            (role) => role._id === roleId
                        )?.title;
                        return z.nativeEnum(UserRoleTitles).parse(title);
                    });

                    console.debug("Sending response...");
                    return {
                        id: user._id,
                        ...user,
                        formioToken,
                        roleTitles,
                        formioTokenExpiration:
                            Date.now() + formioTokenExpirationTime,
                    };
                } catch (e) {
                    console.error("Error: ", e); // log on server
                    if (e instanceof RequestError) {
                        const isInvalidCredentials =
                            e.statusCode >= 400 && e.statusCode < 500;
                        if (isInvalidCredentials) return null;
                    }
                    throw new Error("Unexpected error on server"); // Send generic error to client
                }
            },
        }),
    ],
    jwt: {
        maxAge: 230, // this must be lower that the validity of the formio token
    },
    callbacks: {
        /**
         * Creates the JWT token on the server.
         */
        async jwt({ token, user }) {
            // the user is available only in the first call
            // initial sign in
            if (user) {
                token.user = user;
            } else {
                if (!token.user) throw new Error("No user in token");
                // update the formio token (may throw)
                if (
                    Date.now() - token.user.formioTokenExpiration >
                    3 * 60 * 1000 // 3 minutes have passed
                ) {
                    const newToken = await refreshToken(token.user.formioToken);
                    token.user.formioToken = newToken;
                }
            }

            return token;
        },
        /**
         * Creates the session object on the server, which is sent to the client.
         */
        async session({ session, token }) {
            // make user information available in the session (used in client)
            if (token.user) session.user = token.user;

            return session;
        },
    },
    pages: {
        signIn: "/login",
    },
    debug:
        process.env.NODE_ENV === "development" ||
        process.env.NODE_ENV === "test",
};

/**
 * Wrapper for `getServerSession` so that you don't need to import the `authOptions` in every file.
 *
 * @see https://next-auth.js.org/configuration/nextjs
 */
export const getServerAuthSession = (ctx: {
    req: GetServerSidePropsContext["req"];
    res: GetServerSidePropsContext["res"];
}) => {
    return getServerSession(ctx.req, ctx.res, authOptions);
};
