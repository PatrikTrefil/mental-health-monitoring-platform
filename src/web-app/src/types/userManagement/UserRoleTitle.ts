import UserRoleTitles from "@/constants/userRoleTitles";

export type UserRoleTitle =
    (typeof UserRoleTitles)[keyof typeof UserRoleTitles];
