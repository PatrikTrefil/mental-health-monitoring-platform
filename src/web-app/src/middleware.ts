import { UserRoleTitles } from "@/types/users";
import { NextRequestWithAuth, withAuth } from "next-auth/middleware";
import { NextFetchEvent, NextRequest, NextResponse } from "next/server";
import {
    fetchRoleList,
    getCurrentUser,
    loginAdmin,
} from "./client/formioClient";

export async function middleware(
    req: NextRequestWithAuth,
    event: NextFetchEvent
) {
    if (req.nextUrl.pathname.startsWith("/api/"))
        return await apiMiddleware(req);
    else {
        // this check is here only to handle TS error
        // this should never happen
        if (nextAuthMiddleware instanceof Promise)
            throw new Error("nextAuthMiddleware is a promise");
        // everything else is handled by next auth
        nextAuthMiddleware(req, event);
    }
}

export const config = {
    matcher: ["/uzivatel/:path*", "/zamestnanec/:path*", "/api/:path*"],
};

const nextAuthMiddleware = withAuth(
    async function middleware(req: NextRequestWithAuth) {
        return webpageMiddleware(req);
    },
    {
        callbacks: {
            authorized: ({ token }) => !!token,
        },
    }
);

/**
 * Middleware for checking if user has sufficient privileges to access a webpage.
 * This is not used for API routes.
 */
function webpageMiddleware(req: NextRequestWithAuth) {
    // if accessing /zamestnanec/*, check if user has role ZAMESTNANEC
    // if accessing /uzivatel/*, check if user has role KLIENT_PACIENT
    const hasSufficientPrivileges =
        (req.nextUrl.pathname.startsWith("/zamestnanec/") &&
            req.nextauth.token?.user?.roleTitles.includes(
                UserRoleTitles.ZAMESTNANEC
            )) ||
        (req.nextUrl.pathname.startsWith("/uzivatel/") &&
            req.nextauth.token?.user?.roleTitles.includes(
                UserRoleTitles.KLIENT_PACIENT
            ));

    if (!hasSufficientPrivileges) {
        const url_403 = req.nextUrl.clone();
        url_403.pathname = "/403";
        url_403.searchParams.set("callbackUrl", req.nextUrl.toString());
        return NextResponse.rewrite(url_403);
    }
}

/**
 * Middleware for checking if user has sufficient privileges to access an API route.
 * This is not used for webpages.
 *
 * This middleware does not use next auth. It only expects the formio token to be in the header.
 */
async function apiMiddleware(req: NextRequest) {
    if (req.nextUrl.pathname.startsWith("/api/ukol")) {
        console.log("ahldf");
        const formioToken = req.headers.get("x-jwt-token");
        if (!formioToken)
            return new NextResponse(
                "Missing formio token in header x-jwt-token",
                {
                    status: 401,
                }
            );
        const user = await getCurrentUser(formioToken);
        const adminToken = await loginAdmin(
            process.env.FORMIO_ROOT_EMAIL,
            process.env.FORMIO_ROOT_PASSWORD
        );
        const roleList = await fetchRoleList(adminToken);
        const clientPatientRoleId = roleList.find(
            (r) => r.title === UserRoleTitles.KLIENT_PACIENT
        )?._id;
        if (!clientPatientRoleId)
            return new NextResponse("Something went wrong", { status: 500 });

        if (!user.roles.includes(clientPatientRoleId)) {
            return new NextResponse("Unauthorized", { status: 401 });
        }
    }
}
