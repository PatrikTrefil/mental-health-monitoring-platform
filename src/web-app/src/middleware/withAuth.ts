import {
    fetchRoleList,
    getCurrentUser,
    loginAdmin,
} from "@/client/userManagementClient";
import { UserRoleTitle, UserRoleTitles } from "@/types/users";
import withAuthNextAuth, { NextRequestWithAuth } from "next-auth/middleware";
import { NextRequest, NextResponse } from "next/server";
import { MiddlewareWrapper } from "./types";

/**
 * Middleware for checking if user is authenticated.
 */
const withAuth: MiddlewareWrapper = (middlewareToWrap) => {
    return async (req, event) => {
        console.log("Authenticating in middleware...");
        if (req.nextUrl.pathname.startsWith("/api/")) {
            const errorResponse = await apiMiddleware(req);
            if (errorResponse) return errorResponse;
        } else {
            // this check is here only to handle TS error
            // this should never happen
            if (nextAuthMiddleware instanceof Promise)
                throw new Error("nextAuthMiddleware is a promise");
            // everything else is handled by next auth
            const errorResponse = await nextAuthMiddleware(req, event);
            if (errorResponse) return errorResponse;
        }
        return await middlewareToWrap(req, event);
    };
};
export default withAuth;

const nextAuthMiddleware = withAuthNextAuth(
    async function middleware(req) {
        return webpageMiddleware(req);
    },
    {
        callbacks: {
            authorized: ({ token }) => !!token,
        },
    }
);

/**
 * Middleware for checking if user has sufficient privileges to access a webpage.
 * This is not used for API routes.
 */
function webpageMiddleware(req: NextRequestWithAuth) {
    if (
        req.nextauth?.token?.user &&
        !hasEnoughPrivilegesForWebpage(
            req.nextUrl.pathname,
            req.nextauth.token.user.roleTitles
        )
    ) {
        const url_403 = req.nextUrl.clone();
        url_403.pathname = "/403";
        url_403.searchParams.set("callbackUrl", req.nextUrl.toString());
        return NextResponse.rewrite(url_403);
    }
}

function hasEnoughPrivilegesForWebpage(
    webpagePathname: string,
    roleTitles: UserRoleTitle[]
) {
    if (webpagePathname.startsWith("/zamestnanec/"))
        return (
            roleTitles.includes(UserRoleTitles.SPRAVCE_DOTAZNIKU) ||
            roleTitles.includes(UserRoleTitles.ZADAVATEL_DOTAZNIKU)
        );

    if (webpagePathname.startsWith("/uzivatel/"))
        return roleTitles.includes(UserRoleTitles.KLIENT_PACIENT);
}

/**
 * Middleware for checking if user has sufficient privileges to access an API route.
 * This is not used for webpages.
 *
 * This middleware does not use next auth. It only expects the formio token to be in the header.
 */
async function apiMiddleware(req: NextRequest) {
    if (req.nextUrl.pathname.startsWith("/api/ukol")) {
        const formioToken = req.headers.get("x-jwt-token");
        if (!formioToken)
            return new NextResponse(
                "Missing formio token in header x-jwt-token",
                {
                    status: 401,
                }
            );

        let user: Awaited<ReturnType<typeof getCurrentUser>> | undefined;
        try {
            user = await getCurrentUser(formioToken);
        } catch (e) {
            console.log(e);
            return new NextResponse("Unauthorized", { status: 401 });
        }

        let roleList: Awaited<ReturnType<typeof fetchRoleList>> | undefined;
        try {
            const adminToken = await loginAdmin(
                process.env.FORMIO_ROOT_EMAIL,
                process.env.FORMIO_ROOT_PASSWORD
            );
            roleList = await fetchRoleList(adminToken);
        } catch {
            return new NextResponse("Internal server error", { status: 500 });
        }

        const clientPatientRoleId = roleList.find(
            (r) => r.title === UserRoleTitles.KLIENT_PACIENT
        )?._id;

        if (!clientPatientRoleId)
            return new NextResponse("Internal server error", { status: 500 });

        if (!user.roles.includes(clientPatientRoleId))
            return new NextResponse("Unauthorized", { status: 401 });
    }
}
