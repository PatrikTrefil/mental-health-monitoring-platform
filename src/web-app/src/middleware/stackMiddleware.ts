import { NextMiddlewareWithAuth } from "next-auth/middleware";
import { NextResponse } from "next/server";
import { MiddlewareWrapper } from "./types";

/**
 * Chain multiple middleware functions into one.
 * @param functions - Functions to chain.
 * @returns Chained middleware function.
 */
export function stackMiddlewares(
    functions: MiddlewareWrapper[] = []
): NextMiddlewareWithAuth {
    return stackMiddlewaresInternal(functions, 0);
}

/**
 * Internal function to chain multiple middleware functions into one.
 * @param functions - Functions to chain.
 * @param index - Index of the current function.
 * @returns Chained middleware functions.
 */
function stackMiddlewaresInternal(
    functions: MiddlewareWrapper[] = [],
    index: number
): NextMiddlewareWithAuth {
    // recursively call all middlewares
    const current = functions[index];
    if (current) {
        const next = stackMiddlewaresInternal(functions, index + 1);
        return current(next);
    }
    return () => NextResponse.next();
}
