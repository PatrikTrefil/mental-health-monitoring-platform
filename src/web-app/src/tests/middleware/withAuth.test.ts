import {
    getCurrentUser,
    loadRoles,
    loginAdmin,
} from "@/client/userManagementClient";
import UserRoleTitles from "@/constants/userRoleTitles";
import { stackMiddlewares } from "@/middleware/stackMiddleware";
import withAuth from "@/middleware/withAuth";
import { UserRoleTitle } from "@/types/userManagement/UserRoleTitle";
import { faker } from "@faker-js/faker";
import { NextRequestWithAuth } from "next-auth/middleware";
import { NextRequest } from "next/server";
import { describe, expect, it, vi } from "vitest";

// make tests deterministic
faker.seed(123);

const authMiddleware = stackMiddlewares([withAuth]);

const mockInvalidUserToken = "invalid user token;";
const mockValidUserToken = "valid user token";
const mockValidAdminToken = "valid admin token";

const mockUserRoleId = faker.string.alpha(10);

vi.mock("@/client/userManagementClient", () => ({
    getCurrentUser: vi.fn(async (token: string) => {
        if (token === mockInvalidUserToken) throw new TypeError();
        else if (token === mockValidUserToken) {
            const mockUser: Awaited<ReturnType<typeof getCurrentUser>> = {
                _id: faker.string.uuid(),
                data: { id: faker.string.alpha(5) },
                created: faker.date.recent().toISOString(),
                owner: faker.string.uuid(),
                access: [],
                form: faker.string.uuid(),
                roles: [mockUserRoleId],
                metadata: {},
            };
            return mockUser;
        }
        throw new Error("Unexpected token");
    }),
    loadRoles: vi.fn(),
    loginAdmin: vi.fn(),
}));

const mockNextFetchEvent = {} as any; // There is a problem with standard instantiation of NextFetchEvent

type TestInput = {
    roleTitleToTest: UserRoleTitle;
    pathToTest: string;
};

describe.each<TestInput>([
    {
        roleTitleToTest: UserRoleTitles.ASSIGNEE,
        pathToTest: "http://localhost/uzivatel/prehled",
    },
    {
        roleTitleToTest: UserRoleTitles.FORM_MANAGER,
        pathToTest: "http://localhost/zamestnanec/sprava-formularu",
    },
    {
        roleTitleToTest: UserRoleTitles.ASSIGNER,
        pathToTest: "http://localhost/zamestnanec/sprava-formularu",
    },
])(
    "accessing $roleTitleToTest page",
    ({ roleTitleToTest, pathToTest }: TestInput) => {
        it("should not be possible if no token is provided", async () => {
            const req = new NextRequest(pathToTest) as NextRequestWithAuth;
            // will make it into a valid instance on the next line
            req.nextauth = { token: null };

            const response = await authMiddleware(req, mockNextFetchEvent);
            expect(response?.status).toMatchSnapshot();
        });
        it(`should not be possible if the user is not an ${roleTitleToTest}`, async () => {
            const req = new NextRequest(pathToTest) as NextRequestWithAuth;
            // will make it into a valid instance on the next line
            req.nextauth = {
                token: {
                    user: {
                        roleTitles: [],
                        _id: "123",
                        email: "",
                        created: "",
                        access: [],
                        data: { id: "" },
                        form: "",
                        formioToken: "",
                        metadata: {},
                        owner: "",
                        roles: [],
                        formioTokenExpiration: Infinity,
                    },
                },
            };
            const response = await authMiddleware(req, mockNextFetchEvent);
            expect(response?.status).toMatchSnapshot();
        });
        it(`should be possible if the user is an ${roleTitleToTest}`, async () => {
            const req = new NextRequest(pathToTest) as NextRequestWithAuth; // will make it into a valid instance on the next line
            req.nextauth = {
                token: {
                    user: {
                        roleTitles: [roleTitleToTest],
                        // rest of user object should be irrelevant
                        _id: "",
                        email: "",
                        created: "",
                        access: [],
                        data: { id: "" },
                        form: "",
                        formioToken: "",
                        metadata: {},
                        owner: "",
                        roles: [],
                        formioTokenExpiration: Infinity,
                    },
                },
            };
            const response = await authMiddleware(req, mockNextFetchEvent);
            expect(response?.status).toMatchSnapshot();
        });
    }
);

describe("accessing /api/ukol/*", () => {
    it("should not be possible if no token is provided", async () => {
        const req = new NextRequest(
            "http://localhost/api/ukol/"
        ) as NextRequestWithAuth;
        req.nextauth = { token: null };
        const response = await authMiddleware(req, mockNextFetchEvent);
        expect(response?.status).toMatchInlineSnapshot("401");
    });

    it("should not be possible with invalid token provided", async () => {
        const req = new NextRequest(
            "http://localhost/api/ukol/"
        ) as NextRequestWithAuth;
        req.nextauth = { token: null };
        req.headers.set("x-jwt-token", mockInvalidUserToken);
        const response = await authMiddleware(req, mockNextFetchEvent);
        expect(response?.status).toMatchInlineSnapshot("401");
    });

    it("should be possible with valid user token provided", async () => {
        const req = new NextRequest(
            "http://localhost/api/ukol/"
        ) as NextRequestWithAuth;
        req.nextauth = { token: null };
        req.headers.set("x-jwt-token", mockValidUserToken);

        vi.mocked(loginAdmin).mockImplementationOnce(
            async () => mockValidAdminToken
        );

        vi.mocked(loadRoles).mockImplementationOnce(async () => {
            const mockRoleList: Awaited<ReturnType<typeof loadRoles>> = [
                {
                    _id: mockUserRoleId,
                    title: UserRoleTitles.ASSIGNEE,
                },
            ];
            return mockRoleList;
        });

        const response = await authMiddleware(req, mockNextFetchEvent);
        expect(response?.status).toMatchInlineSnapshot("200");
    });
});
