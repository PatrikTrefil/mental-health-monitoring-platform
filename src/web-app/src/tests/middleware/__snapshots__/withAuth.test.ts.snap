// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`accessing 'Klient/Pacient' page > should be possible if the user is an Klient/Pacient 1`] = `307`;

exports[`accessing 'Klient/Pacient' page > should not be possible if no token is provided 1`] = `307`;

exports[`accessing 'Klient/Pacient' page > should not be possible if the user is not an Klient/Pacient 1`] = `307`;

exports[`accessing 'Správce dotazníků' page > should be possible if the user is an Správce dotazníků 1`] = `307`;

exports[`accessing 'Správce dotazníků' page > should not be possible if no token is provided 1`] = `307`;

exports[`accessing 'Správce dotazníků' page > should not be possible if the user is not an Správce dotazníků 1`] = `307`;

exports[`accessing 'Zadavatel dotazníků' page > should be possible if the user is an Zadavatel dotazníků 1`] = `307`;

exports[`accessing 'Zadavatel dotazníků' page > should not be possible if no token is provided 1`] = `307`;

exports[`accessing 'Zadavatel dotazníků' page > should not be possible if the user is not an Zadavatel dotazníků 1`] = `307`;

exports[`accessing 'Zamestnanec' page > should be possible if the user is an Zamestnanec 1`] = `307`;

exports[`accessing 'Zamestnanec' page > should not be possible if no token is provided 1`] = `307`;

exports[`accessing 'Zamestnanec' page > should not be possible if the user is not an Zamestnanec 1`] = `307`;
